notebooks/14_regression/index.md

```diff
@@ -0,0 +1,222 @@
+    \end{bmatrix}
+$$
+where $x^{(i)}$ is the i'th data point. 
+
+## Loss Function
+After defining $f$, we need to find the best function. By defining a loss function, we can try to minimize loss by changing $w$ in the main function. Assuming $L$ as our loss function, best $f$ will be
+
+$$
+    \hat{w} = argmin_w L(y_w, \hat{y}) \rightarrow f_{best}(x) = f_{\hat{w}}(x) 
+$$
+
+where $\hat{y}$ is the given number for each data point.
+
+### Mean Squared Error (MSE)
+The main loss function we use is mean squared error. It's defined as 
+
+$$
+    MSE(y_w, \hat{y}) = \frac{1}{2} \Sigma_{i=1}^m \left[ y_w^{(i)} - \hat{y}^{(i)} \right]^2 .
+$$
+The main reason for using this function is that we can calculate gradients easily. So we can use gradient descent to find $\hat{w}$.
+
+## Finding $\hat{w}$
+
+### Gradient Descent
+We want to use gradient descent to find $\hat{w}$. First, we need to calculate $\nabla_w L(y_w, \hat{y})$ because it's used in the gradient descent method. The partial derivatives for MSE are:
+
+$$
+    \frac{\partial MSE}{\partial w_j} = - \Sigma_{i = 1}^m x_j^{(i)} \left[ y_w^{(i)} - \hat{y}^{(i)} \right]
+$$
+
+So for gradinet, we have:
+
+$$
+    \nabla_w MSE = \begin{bmatrix}
+        \frac{\partial MSE}{\partial w_0} \\ \vdots \\ \frac{\partial MSE}{\partial w_n}
+    \end{bmatrix}
+$$
+
+Now, we can find $\hat{w}$ by the gradient descent algorithm as
+
+$$
+    w^{(i+1)} = w^{(i)} - \eta \nabla_{w^{(i)}} L
+$$
+
+where $\eta$ is the learning rate.
+
+### Normal Equation
+If we define
+
+$$
+    X = \begin{bmatrix}
+        x^{(1)} \\ \vdots \\ x^{(m)}
+    \end{bmatrix}
+$$
+
+and solve the equation $\nabla_w MSE = 0$, we get the normal equation. It's defined as
+$$
+    \hat{w} = (X^TX)^{-1} X^T \hat{y}.
+$$
+However, for using this equation, our features must be linearly independent. Otherwise, $(X^TX)^{-1}$ is not defined. In that case, we can use the pseudo inverse of $X^TX$ instead of the $(X^TX)^{-1}$. Since the calculation of inverse is computationally inefficient, we usually prefer using gradient descent. 
+
+# 2 - Learning Curves Using Polynomials
+
+In this section, we want to find the best $P(x)$ where x is a real number and $P$ is an n'th degree polynomial.
+
+## Reduction to Linear Regression
+
+We can define 
+$$
+    z = \begin{bmatrix}
+        x^0 \\ x^1 \\ \vdots \\ x^n
+    \end{bmatrix},
+$$
+then use linear regression to find $f_w(z) = w^T z$. From definition, we know that $P(x) = f_w(z)$. However, if $n$ is too large, overfitting might happen since we are getting closer and closer to interpolation.
+
+## Overfitting
+To prevent overfitting, we must try to define $n$ optimally. Since $n$ is a hyperparameter here, we can use validation set.
+
+The example below demonstrates how $n$ can change the curve we find and when overfitting happens. The green curve is our goal. The red curve is the M'th degree polynomial using this method.
+
+![M = 1](images/1.png)
```

Comment:
Show figures in markdown instead of hyperlinking

Reply:
Using base64 of images?

Reply:
The way you did it is correct